File: userspace/coreutils/prompt.c
/* /bin/PROMPT — ion shell prompt command.
 * Ion executes this as a command to generate the prompt string.
 * Output is used verbatim as the prompt (ANSI escapes supported).
 */
#include <unistd.h>
#include <string.h>
#include <sys/types.h>

/* Convert uid to decimal string, return length. */
static int uid_to_str(uid_t uid, char *buf, int bufsz) {
    if (uid == 0) { buf[0] = '0'; return 1; }
    char tmp[16];
    int pos = 0;
    while (uid > 0 && pos < (int)sizeof(tmp)) {
        tmp[pos++] = '0' + (uid % 10);
        uid /= 10;
    }
    if (pos > bufsz) pos = bufsz;
    for (int i = 0; i < pos; i++) buf[i] = tmp[pos - 1 - i];
    return pos;
}

int main(void) {
    /* Build prompt: \x1b[94m<user>\x1b[0m@\x1b[96mm3os\x1b[0m:<cwd>$ */
    char prompt[512];
    int len = 0;

    /* Username from $USER, fallback to uid. */
    const char *user = NULL;
    /* Read USER env var via /proc-style or getenv - use extern. */
    extern char **environ;
    if (environ) {
        for (char **e = environ; *e; e++) {
            if ((*e)[0]=='U' && (*e)[1]=='S' && (*e)[2]=='E' && (*e)[3]=='R' && (*e)[4]=='=') {
                user = *e + 5;
                break;
            }
        }
    }

    /* Light blue username. */
    memcpy(prompt + len, "\x1b[94m", 5); len += 5;
    if (user && *user) {
        int ul = strlen(user);
        if (ul > 32) ul = 32;
        memcpy(prompt + len, user, ul); len += ul;
    } else {
        /* Fallback: show uid. */
        len += uid_to_str(getuid(), prompt + len, 16);
    }
    memcpy(prompt + len, "\x1b[0m", 4); len += 4;

    /* @hostname in cyan. */
    prompt[len++] = '@';
    memcpy(prompt + len, "\x1b[96m", 5); len += 5;
    memcpy(prompt + len, "m3os", 4); len += 4;
    memcpy(prompt + len, "\x1b[0m", 4); len += 4;

    /* :cwd */
    prompt[len++] = ':';
    char cwd[256];
    if (getcwd(cwd, sizeof(cwd))) {
        int cl = strlen(cwd);
        if (cl > 128) cl = 128;
        memcpy(prompt + len, cwd, cl); len += cl;
    }

    /* # for root, $ for others. */
    if (getuid() == 0)
        memcpy(prompt + len, "# ", 2);
    else
        memcpy(prompt + len, "$ ", 2);
    len += 2;

    write(1, prompt, len);
    return 0;
}

File: userspace/coreutils-rs/src/prompt.rs
//! PROMPT — ion shell prompt generator.
//!
//! Output: `\x1b[94m<user>\x1b[0m@\x1b[96mm3os\x1b[0m:<cwd># ` (root)
//!     or: `\x1b[94m<user>\x1b[0m@\x1b[96mm3os\x1b[0m:<cwd>$ ` (non-root)
#![no_std]
#![no_main]

use syscall_lib::{STDOUT_FILENO, getcwd, getuid, write};

syscall_lib::entry_point_with_env!(main);

fn main(_args: &[&str], env: &[&str]) -> i32 {
    let mut prompt = [0u8; 512];
    let mut len = 0;

    // Find $USER in environment.
    let mut user: Option<&str> = None;
    for e in env {
        if let Some(val) = e.strip_prefix("USER=") {
            user = Some(val);
            break;
        }
    }

    // Light blue username.
    len += copy(&mut prompt[len..], b"\x1b[94m");
    if let Some(u) = user {
        let ub = u.as_bytes();
        let ul = ub.len().min(32);
        len += copy(&mut prompt[len..], &ub[..ul]);
    } else {
        // Fallback: show uid as decimal.
        len += write_uint_to(&mut prompt[len..], getuid() as u64);
    }
    len += copy(&mut prompt[len..], b"\x1b[0m");

    // @hostname in cyan.
    prompt[len] = b'@';
    len += 1;
    len += copy(&mut prompt[len..], b"\x1b[96m");
    len += copy(&mut prompt[len..], b"m3os");
    len += copy(&mut prompt[len..], b"\x1b[0m");

    // :cwd
    prompt[len] = b':';
    len += 1;
    let mut cwd = [0u8; 256];
    let ret = getcwd(&mut cwd);
    if ret >= 0 {
        let cwd_len = cwd.iter().position(|&b| b == 0).unwrap_or(ret as usize);
        let cl = cwd_len.min(128);
        len += copy(&mut prompt[len..], &cwd[..cl]);
    }

    // # for root, $ for others.
    if getuid() == 0 {
        len += copy(&mut prompt[len..], b"# ");
    } else {
        len += copy(&mut prompt[len..], b"$ ");
    }

    let _ = write(STDOUT_FILENO, &prompt[..len]);
    0
}

fn copy(dst: &mut [u8], src: &[u8]) -> usize {
    let n = src.len().min(dst.len());
    dst[..n].copy_from_slice(&src[..n]);
    n
}

fn write_uint_to(dst: &mut [u8], mut v: u64) -> usize {
    if v == 0 {
        if !dst.is_empty() {
            dst[0] = b'0';
        }
        return 1;
    }
    let mut tmp = [0u8; 20];
    let mut pos = 20;
    while v > 0 {
        pos -= 1;
        tmp[pos] = b'0' + (v % 10) as u8;
        v /= 10;
    }
    let digits = &tmp[pos..];
    let n = digits.len().min(dst.len());
    dst[..n].copy_from_slice(&digits[..n]);
    n
}

#[panic_handler]
fn panic(_: &core::panic::PanicInfo) -> ! {
    syscall_lib::exit(101)
}
